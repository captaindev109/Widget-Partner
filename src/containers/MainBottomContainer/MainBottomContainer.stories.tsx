import React from 'react';
import { Meta, Story } from '@storybook/react';
import styled from 'styled-components';
import { decorators } from '@/../.storybook/preview';
import Wrapper from '@/generics/Wrapper';
import MainBottomContainer, { MainBottomContainerProps } from './MainBottomContainer';

// A wrapper to position the component
const StoryWrapper = styled(Wrapper)``;

// Name the folder to either generics or components
const meta: Meta = {
  title: 'Components/MainBottomContainer',
  component: MainBottomContainer,
  decorators,
};

export default meta;

// Showcasing the component
const Template: Story<MainBottomContainerProps> = (args) => (
  <StoryWrapper>
    <MainBottomContainer {...args} />
    <MainBottomContainer {...args} />
  </StoryWrapper>
);

export const DefaultMainBottomContainer = Template.bind({});

DefaultMainBottomContainer.args = {
  children: 'Badge',
};
