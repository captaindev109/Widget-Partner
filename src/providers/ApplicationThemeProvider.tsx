import { ThemeProvider, DefaultTheme } from 'styled-components';
import GlobalStyles from '@/styles/Global';
import React from 'react';
import { ChildrenReturnType } from '@/types/types';

interface ApplicationThemeProviderProps {
  children?: ChildrenReturnType;
  theme: DefaultTheme;
}

const ApplicationThemeProvider = ({ children, theme }: ApplicationThemeProviderProps) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};

export default ApplicationThemeProvider;
