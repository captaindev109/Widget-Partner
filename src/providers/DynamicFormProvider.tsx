import React from 'react';
import { ChildrenReturnType } from '@/types/types';
import { FormProvider, useForm } from 'react-hook-form';

interface DynamicFormProviderProps {
  children?: ChildrenReturnType;
}

const useDynamicRootForm = () => {
  const methods = useForm({ mode: 'onChange' });

  return { methods };
};

export const DynamicFormProvider = ({ children }: DynamicFormProviderProps) => {
  const { methods } = useDynamicRootForm();

  return <FormProvider {...methods}>{children}</FormProvider>;
};
