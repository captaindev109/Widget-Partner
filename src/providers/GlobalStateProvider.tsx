import { mainMachine } from '@/machines/courseMachine';
import headerMachine from '@/machines/headerMachine';
import footerMachine from '@/machines/footerMachine';
import React, { createContext } from 'react';
import { useInterpret } from '@xstate/react';
import { ActorRefFrom } from 'xstate';
import { storeMachine } from '@/machines/storeMachine';

interface GlobalStateContextProps {
  mainService: ActorRefFrom<typeof mainMachine>;
  headerService: ActorRefFrom<typeof headerMachine>;
  footerService: ActorRefFrom<typeof footerMachine>;
  storeService: ActorRefFrom<typeof storeMachine>;
}

export const GlobalStateContext = createContext({} as GlobalStateContextProps);

interface GlobalStateProviderProps {
  children?: React.ReactNode;
}

const useDevTool = process.env.NODE_ENV === 'development';

export const GlobalStateProvider = ({ children }: GlobalStateProviderProps) => {
  const services = {
    mainService: useInterpret(mainMachine, {
      devTools: useDevTool,
    }),
    headerService: useInterpret(headerMachine, {
      devTools: useDevTool,
    }),
    footerService: useInterpret(footerMachine, {
      devTools: useDevTool,
    }),
    storeService: useInterpret(storeMachine, {
      devTools: useDevTool,
    }),
  };

  return <GlobalStateContext.Provider value={services}>{children}</GlobalStateContext.Provider>;
};
