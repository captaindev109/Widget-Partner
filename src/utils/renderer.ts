import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
