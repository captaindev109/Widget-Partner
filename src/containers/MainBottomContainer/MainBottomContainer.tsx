import { FC, ReactNode, useRef, useEffect } from 'react';
import styled from 'styled-components';
import Card from '@/generics/Card';
import { BasicAny } from '@/types/types';
import { device } from '@/utils/constants';
import useFooter from '@/hooks/useFooter';

// A basic Interface or type is required for every each component
export interface MainBottomContainerProps {
  children?: ReactNode;
  padding?: boolean;
  shadow?: boolean;
}

// The styled component is using composition to inherit the base from Card component
const StyledMainBottomContainer = styled(Card)<MainBottomContainerProps>`
  width: 100%;
  position: absolute;
  bottom: 0;
  background: white;
  ${(props) => props.padding === true && 'padding: 16px'};
  ${(props) =>
    props.padding === true && 'box-shadow: 0px -4px 8px rgba(0, 0, 0, 0.05), 0px -8px 16px rgba(0, 0, 0, 0.05)'};
  @media ${device.tablet} {
    position: relative;
    padding: 30px 16px;
    background: transparent;
    box-shadow: none;
  }
`;

//The main component with dynamic rendering and state
const MainBottomContainer: FC<MainBottomContainerProps> = ({ children, padding = true, shadow = true }: BasicAny) => {
  const { setHeight } = useFooter();

  const ref = useRef<HTMLDivElement>(null);
  useEffect(() => {
    if (ref.current && ref.current.clientHeight) {
      setHeight(ref.current.clientHeight);
    } else {
      setHeight(0);
    }
  }, [children, ref, setHeight]);
  return (
    <StyledMainBottomContainer padding={padding} shadow={shadow} data-testid="MainBottomContainer" ref={ref}>
      {children}
    </StyledMainBottomContainer>
  );
};

export default MainBottomContainer;
