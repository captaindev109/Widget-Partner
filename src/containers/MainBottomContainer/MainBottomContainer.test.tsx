/**
 *
 *  ********** Unit & Snapshot Testing *********
 *
 * Unit test is specifiic for each functionliaty of the component
 * Snapshot test is to track Over All changes in the code since the first snap
 **/

import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import MainBottomContainer from './MainBottomContainer';
import ApplicationThemeProvider from '@/providers/ApplicationThemeProvider';
import defaultTheme from '@/theme/theme';
import { GlobalStateProvider } from '@/providers/GlobalStateProvider';

const MockComponent: React.FC = () => {
  return (
    <GlobalStateProvider>
      <ApplicationThemeProvider theme={defaultTheme}>
        <MainBottomContainer />
      </ApplicationThemeProvider>
    </GlobalStateProvider>
  );
};

// describe function is a better way to organize our test cases and describe which component we are testing
describe('<MainBottomContainer /> Test Cases', () => {
  // this test is a must for each component, any rendering issues with the component can be found
  test('it renders without crashing, all is well', () => {
    // render basically prints and exposes the component in nodejs environemt, it returns an object...
    render(<MockComponent />);
  });

  it('Snapshot test should pass', () => {
    const snap = renderer.create(<MockComponent />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
