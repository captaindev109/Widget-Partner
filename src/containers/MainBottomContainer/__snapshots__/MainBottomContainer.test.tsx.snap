// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainBottomContainer /> Test Cases Snapshot test should pass 1`] = `
<div
  className="Card-sc-fb64gk-0 MainBottomContainer__StyledMainBottomContainer-sc-m5h2jy-0 jpqWTd cfxCCa"
  data-testid="MainBottomContainer"
/>
`;
